src/main/scala/latis/service/hapi/InfoService.scala

```diff
@@ -17,27 +17,37 @@ class InfoService[F[_]: Concurrent](alg: InfoAlgebra[F]) extends Http4sDsl[F] {
   val service: HttpRoutes[F] =
     HttpRoutes.of[F] {
       case GET -> Root / "info"
-          :? IdMatcher(id)
+          :? DatasetMatcher(ds)
+          +& IdMatcher(id)
           +& ParamMatcher(params) =>
-        val ps = params.map(_.distinct)
-        alg.getMetadata(id, ps).leftMap {
-          case UnknownId(_)          => Status.`1406`
-          case UnknownParam(_)       => Status.`1407`
-          case err @ MetadataError(_)      => logger.info(err.toString); Status.`1501`
-          case err @ UnsupportedDataset(_) => logger.info(err.toString); Status.`1501`
-        }.fold(
-          err => err match {
-            case Status.`1406` | Status.`1407` =>
-              logger.info(err.message)
-              NotFound(HapiError(err).asJson)
-            case _ =>
-              logger.info(err.message)
-              InternalServerError(HapiError(err).asJson)
+        // Allow both "dataset" or "id" for backwards compatibility
+        Either.fromOption(ds <+> id, Status.`1400`).fold(
+          err => {
+            logger.info(err.message)
+            InternalServerError(HapiError(Status.`1400`).asJson)
           },
```

Comment:
This works but it feels kinda funny that I'm folding twice. I tried to combine this fold with the rest of the original code (everything in the second half of this fold) such that there'd only be one fold, but it wouldn't work.

Reply:
I think it would be cleaner to get the dataset ID as a `EitherT[F, Status, String]` (take a look at the `EitherT` companion object) and then combine that with the rest using a `flatMap` or in a for comprehension.

Reply:
Sorry can you say more about how to do this? I looked at the companion object and I'm guessing I'd use either `.fromOptionF` or `.fromOptionM`, but it's not obvious which or how.

Reply:
You can replace `Either.fromOption(...)` with `EitherT.fromOption[F](...)`. The `fromOptionF` and `fromOptionM` methods are for when what we're lifting into `EitherT` is already in `F`, but it's not in our case.

Once you have that, you could rewrite to look something like this:

```scala
(for {
  dataset  <- EitherT.fromOption[F](...)
  metadata <- alg.getMetadata(...).leftMap {
    // InfoError => Status
    case ...
  }
} yield metadata).foldF(
  err => ...,
  res => ...
)
```

I didn't notice this the first time, but instead of `InternalServerError` we should use `BadRequest`.

The code as it was originally written used a `fold` followed by a `flatten`, but when I was thinking about this I found out that there's a `foldF` method that will do what we want without the `flatten`.

Reply:
Sweet thanks, that's satisfying. I'd written something nearly identical last week but the types never worked out because I was using plain `Either` instead of `EitherT`. Happy to see it work.